### DISCUSSION TITLE
在哪儿配置机器人在线设备

### DISCUSSION THREAD

**Discussion by fhh-coder** [simple-robot/Discussions#81] [Q&A] [answered]
### 涉及的编程语言

Java

### 涉及的组件库

mirai (QQ)

### 疑问描述

现在是simbot显示的是手机登录，手机qq登录，会把simbot的登录挤下线。
1. 请问怎么修改simbot的登录显示？
2. 如果要动态登录（注册机器人），请问怎么操作？比如给在线的机器人输入qq和密码，或者用二维码的形式，进行登录。

**Comment by ForteScarlet** [comment, marked as answer]
> 怎么修改simbot的登录显示？

bot使用的登录协议参考 [mirai组件bot配置](https://component-mirai.simbot.forte.love/docs/bot-config) 中的 `config.protocol`

> 怎么动态注册BOT

动态登录可以获取到组件对应的`BotManager` 后使用特定的 `register` 注册。在mirai组件中，其类型为 `MiraiBotManager`，可以在 `Application.botManagers` 中迭代并寻找。

```java
var managers = application.getBotManagers();

for(var manager : managers) {
   if(manager instanceof MiraiBotManager miraiBotManager) {
     MiraiBot bot = manager.register(...);
     bot.startBlocking(); // or startAsync
      // 或其他操作 ......
   }
}

```

其中，在SpringBoot环境中 `Application` 实例可以自动注入。

**Reply by zhihui-Yu** [reply]
![image](https://user-images.githubusercontent.com/54210717/220071832-67affaea-6784-460c-b8b0-27831c3817c6.png)
这里是不是写错了呢

**Reply by ForteScarlet** [reply]
啊是，应该是 `MiraiBotManager` ，手机打的字不太准

**Comment by fhh-coder** [comment]
我试了一下，使用配置bot.json的方式，可以登录。但是使用动态的注册方法，需要验证码，而且验证完了，会提示“版本过低”的提示，造成无法成功登录。

**Reply by ForteScarlet** [reply]
使用注册的方式需要你自行提供配置类属性，包括但不限于 `device.json`，`cache` 目录 （例如处理 https://github.com/mamoe/mirai/issues/2475 中的问题）等相关内容。

**Comment by chenwenhang** [comment]
@ForteScarlet 请教下咱们现在对动态启动机器人有新的使用方式么？
我这边想实现的是允许各租户自己配置机器人比如 kook 机器人的 token，然后保存到数据库里。可以支持动态触发读取数据库的机器人配置并启动监听～

**Reply by ForteScarlet** [reply]
新的方式指？目前动态注册还是通过 `BotManager.register` (比如找到 `KookBotManager` 后使用 `register`) 动态注册bot，存在数据库你也可以查询、遍历并动态注册。

**Reply by chenwenhang** [reply]
> 新的方式指？目前动态注册还是通过 `BotManager.register` (比如找到 `KookBotManager` 后使用 `register`) 动态注册bot，存在数据库你也可以查询、遍历并动态注册。

好的了解，那我先按照这个方式尝试看看，感谢～
